components/calendar/src/persian.rs

```diff
@@ -0,0 +1,186 @@
+use crate::{types, Calendar, Date, DateDuration, DateDurationUnit, Iso};
+use tinystr::tinystr;
+
+#[derive(Copy, Clone, Debug, Default)]
+/// The Persian Calendar
+pub struct Persian;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct PersianDateInner {
+    year: i32,
+    month: u8,
+    day: u8,
+}
+
+impl Persian {
+    fn is_leap_year(year: i32) -> bool {
+        (year * 25 + 11) % 33 < 8
+    }
+
+    fn day_of_year(date: &PersianDateInner) -> i32 {
+        Persian::MONTH_PREFIX_SUM[date.month as usize - 1] + date.day as i32 - 1
+    }
+
+    const EPOCH: i32 = 226895;
+    const MONTH_PREFIX_SUM: [i32; 12] = [0, 31, 62, 93, 124, 155, 186, 216, 246, 276, 306, 336];
+}
+
+impl Calendar for Persian {
+    type DateInner = PersianDateInner;
+
+    // Source: https://github.com/hhstechgroup/icu4j/blob/master/main/classes/core/src/com/ibm/icu/util/PersianCalendar.java
+    fn date_from_iso(&self, iso: Date<Iso>) -> Self::DateInner {
```

Comment:
issue: This code is also based on https://github.com/EdReingold/calendar-code2/blob/main/calendar.l, could you reference that directly and make the code look more like the source?

Reply:
According to the comment on ICU4J

> The exact locale to be used for the Iranian calendar starts to make a
difference at around 2090 CE.  The specific arithmetic method implemented
here, commonly known as the 33-year cycle rule, matches the astronomical
calendar at least for the whole period that the calendar has been both
well-defined and official, from 1925 to around 2090 CE.  The other
commonly known algorithm, the 2820-year cycle, has been incorrectly
designed to follow the tropical year instead of the spring equinoctial
year, and fails to match the astronomical one as early as 2025 CE.

Seems lisp code is using 2820-year cycles, so we should keep the java version? Not that I'm expert on this topic, I just remembered when saw lisp code that 2820 is familiar, then I found this comment.

Reply:
I think that's the arithmetic-persian calendar, not the persian calendar. The code supports both, and the corresponding book does have similar notes about the arithmetic ones